components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,329 @@
+                           *spender_address),
+        std::make_tuple(block_number, mojom::AllowanceInfo::New(
+                                          *contract_address, *approver_address,
+                                          *spender_address, *amount)));
+  }
+}
+
+void EthAllowanceManager::OnGetAllowances(
+    base::OnceCallback<
+        void(base::flat_map<
+             std::string,
+             std::tuple<uint256_t, std::vector<mojom::AllowanceInfoPtr>>>)>
+        barrier_callback,
+    const std::string& chain_id,
+    const std::string& hex_account_address,
+    [[maybe_unused]] const std::vector<Log>& logs,
+    base::Value rawlogs,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  base::flat_map<std::string,
+                 std::tuple<uint256_t, std::vector<mojom::AllowanceInfoPtr>>>
+      allowances;
+
+  if (error != mojom::ProviderError::kSuccess) {
+    std::move(barrier_callback).Run(std::move(allowances));
+    return;
+  }
+
+  std::vector<Log> logs_tmp{logs};
+  std::sort(logs_tmp.begin(), logs_tmp.end(),
+            [](const Log& first, const Log& second) -> bool {
+              if (first.block_number < second.block_number) {
+                return true;
+              }
+              if (first.block_number == second.block_number) {
+                return first.log_index < second.log_index;
+              }
+              return false;
+            });
+
+  // Collection of the latest allowances per contract & spender.
+  base::flat_map<std::string,
+                 std::tuple<brave_wallet::uint256_t, mojom::AllowanceInfoPtr>>
+      allowances_map;
+
+  auto allowance_map_key = [](const std::string& contract_address,
+                              const std::string& approver_addr,
+                              const std::string& spender_address) {
+    return base::JoinString({contract_address, approver_addr, spender_address},
+                            "_");
+  };
+  // Put cashed data into the map first.
+  LoadCachedAllowances(chain_id, hex_account_address, allowances_map);
+  for (const auto& log_item : logs_tmp) {
+    // Skip pending logs.
+    if (log_item.block_number == 0) {
+      continue;
+    }
+    if (log_item.topics.size() < 3) {
+      continue;
+    }
+
+    allowances_map.insert_or_assign(
+        allowance_map_key(log_item.address, log_item.topics[1],
+                          log_item.topics[2]),
+        std::make_tuple(
+            log_item.block_number,
+            mojom::AllowanceInfo::New(log_item.address, log_item.topics[1],
+                                      log_item.topics[2], log_item.data)));
+  }
+  std::vector<mojom::AllowanceInfoPtr> allowances_found;
+  uint256_t max_block_number{0};
+  for (auto& [contract_address, allowance_tuple] : allowances_map) {
+    uint256_t parsed_amount{0};
+    if (HexValueToUint256(std::get<1>(allowance_tuple)->amount,
+                          &parsed_amount) &&
+        parsed_amount > 0) {
```

Comment:
Let's just not insert it in the first place when amount = 0 in the `for (const auto& log_item : logs_tmp)` loop above to minimize the size of allowance_map. We can do `CHECK(parsed_amount > 0);` here to be safe.